src/server/engine_shard_set.cc

```diff
@@ -56,6 +71,95 @@ EngineShard::Stats& EngineShard::Stats::operator+=(const EngineShard::Stats& o)
   return *this;
 }
 
+// This function checks 3 things:
+// 1. Don't try memory fragmentation if we don't use "enough" memory (control by
+// mem_defrag_threshold flag)
+// 2. That we had change in memory usage - to prevent endless loop - out of scope for now
+// 3. in case the above is OK, make sure that we have a "gap" between usage and commited memory
+// (control by commit_use_threshold flag)
+bool EngineShard::DefragTaskState::IsRequired() {
+  const double threshold_mem = max_memory_limit * GetFlag(FLAGS_mem_defrag_threshold);
+  const double commit_use_threshold = GetFlag(FLAGS_commit_use_threshold);
+
+  if (cursor > 0) {
+    return true;
+  }
+
+  uint64_t commited = GetMallocCurrentCommitted();
+  uint64_t mem_in_use = used_mem_current.load(memory_order_relaxed);
+
+  // we want to make sure that we are not running this to many times - i.e.
+  // if there was no change to the memory, don't run this
+  if (threshold_mem < commited && mem_in_use != 0 &&
+      (uint64_t(mem_in_use * commit_use_threshold) < commited)) {
+    // we have way more commited then actual usage
+    return true;
+  }
+
+  return false;
+}
+
+// for now this does nothing
+bool EngineShard::DoDefrag() {
+  constexpr size_t CURSOR_COUNT_RELATION = 50;
+  const float threshold = GetFlag(FLAGS_mem_utilization_threshold);
+
+  auto& slice = db_slice();
+  DCHECK(slice.IsDbValid(DEFAULT_DB_INDEX));
+  unsigned cnt = 0;
+  std::size_t max = slice.DbSize(DEFAULT_DB_INDEX) / CURSOR_COUNT_RELATION;
+
+  PrimeTable::Cursor cur = defrag_state_.cursor;
+  auto [prime_table, expire_table] = slice.GetTables(DEFAULT_DB_INDEX);
+  uint64_t count = 0;
```

Comment:
Maybe call cnt and count distinctively

Reply:
I think these are bad names (they were first for internal debugging, then I start using them for the stats). I think they need better treatment.

Reply:
this will wait for the "next" PR - with the implementation of the internals of this task